### PR TITLE
profiling: add Task/GC/Inference/Codegen signposts for Tracy

### PR DESCRIPTION
**Pull request by topolarity**
This adds a variety of metadata to send to Tracy during profiling, in particular method instance signatures, the Heap size, GC Mark/Sweep/Stop-the-World, and a couple of other small pieces of relevant metadata.

It also adds full support for Tasks, tracking these as "Fibers" on the Tracy side:

Check out an example trace here: https://topolarity.github.io/tracy-traces/tracy-experimental/

image

##### Loose ends

Before this is ready for final review, I still need to:
 - [x] Finalize the JL_TIMING API changes
 - [x] Resolve incompatibility between `TRACY_ON_DEMAND` and Tracy's Fiber support

### CODE REVIEW

File: src/support/ios.h [hunk 1]
```diff
@@ -19,13 +19,13 @@ extern "C" {
 typedef enum { bm_none=1000, bm_line, bm_block, bm_mem } bufmode_t;
 typedef enum { bst_none, bst_rd, bst_wr } bufstate_t;
 
-#define IOS_INLSIZE 54
+#define IOS_INLSIZE 83
```
Comment (gbaraldi) [inline src/support/ios.h:22]: Why was this changed?

File: src/jitlayers.cpp [hunk 2]
```diff
@@ -248,9 +248,12 @@ static jl_callptr_t _jl_compile_codeinst(
         IndirectCodeinsts += emitted.size() - 1;
     }
 
+    size_t i = 0;
     for (auto &def : emitted) {
         jl_code_instance_t *this_code = def.first;
-        jl_timing_show_func_sig(this_code->def->specTypes, JL_TIMING_CURRENT_BLOCK);
+        if (i < 10)
```
Comment (kpamnany) [inline src/jitlayers.cpp:254]: Maybe make this controllable with an optional environment variable?

File: src/jlapi.c [hunk 3]
```diff
@@ -680,6 +684,11 @@ static void rr_detach_teleport(void) {
 
 JL_DLLEXPORT int jl_repl_entrypoint(int argc, char *argv[])
 {
+#ifdef USE_TRACY
+    if (getenv("WAIT_FOR_TRACY"))
+        while (!TracyCIsConnected) ; // Wait for connection
```
Comment (kpamnany) [inline src/jlapi.c:689]: Better a `jl_cpu_pause()` rather than just spinning.

File: src/julia.h [hunk 4]
```diff
@@ -1934,6 +1934,10 @@ typedef struct _jl_task_t {
     uint16_t priority;
 
 // hidden state:
+
+#ifdef USE_TRACY
+    const char *name;
+#endif
```
Comment (kpamnany) [inline src/julia.h:1940]: Mentioned on the call, adding this note so it isn't lost: would be great to have a `name` irrespective of Tracy.

**Review by kpamnany — APPROVED** [review]
Looks fantastic!

Another useful point to instrument might be thread sleep/wakeup; search for `PROBE` in `partr.c`.

File: src/jlapi.c [hunk 5]
```diff
@@ -680,6 +684,11 @@ static void rr_detach_teleport(void) {
 
 JL_DLLEXPORT int jl_repl_entrypoint(int argc, char *argv[])
 {
+#ifdef USE_TRACY
```
Comment (staticfloat) [inline src/jlapi.c:687]: I think it would be very useful to have an environment variable that allows changing `jl_timing_enable_mask` before we're running Julia user code.  Something like `JULIA_TIMING_SUBSYSTEMS=INFERENCE,GC,METHOD_MATCH,-SYSIMG_LOAD` which then gets decoded into turning on the first three elements and turning the last one off (relative to the default).  That way, we can have sensible defaults for `jl_timing_enable_mask` without limiting flexibility.

File: src/jitlayers.cpp (hunk 2 above)
Comment (topolarity) [reply src/jitlayers.cpp:254]: Any recommendations about what to call it?

How about something fairly coarse like `JULIA_VERBOSE_TIMING_METADATA=1` that would turn off limits like this in general?

File: src/jlapi.c (hunk 5 above)
Comment (topolarity) [reply src/jlapi.c:687]: A great idea! Let me see what I can whip up.

File: src/jitlayers.cpp (hunk 2 above)
Comment (staticfloat) [reply src/jitlayers.cpp:254]: JULIA_TIMING_METADATA_PRINT_LIMIT, which defaults to 10?

**Comment by topolarity** [comment]
> Another useful point to instrument might be thread sleep/wakeup; search for PROBE in partr.c.

Great idea - I've added that to my list for the follow-up PR (which is targeting locks, etc.)

One other detail that I noticed today is that `JULIA_WAIT_FOR_TRACY` interacts poorly with precompilation. The spawned Julia processes will see the environment variable, also wait for a Tracy connection on start-up, and will make no progress.

Any ideas about how we should handle that?

**Comment by staticfloat** [comment]
I would just delete that environment variable by adding a mapping to `nothing` [here](https://github.com/JuliaLang/julia/blob/ff7b8eb00bf887f20bf57fb7e53be0070a242c07/base/loading.jl#L2104).

**Review by staticfloat — APPROVED** [review]

**Comment by carstenbauer** [comment]
Great stuff! However, I noticed that this PR doesn't add any documentation for the new features. I think that would be a very valuable addition.
